--- parsec-3.0-origin/ext/splash2x/apps/fmm/src/fmm.C	2021-11-09 18:09:12.293454150 -0500
+++ parsec-3.0/ext/splash2x/apps/fmm/src/fmm.C	2021-11-10 23:07:49.608397787 -0500
@@ -361,16 +361,16 @@
 	 exit(-1);
       }
    }
-
-   Total_Particles = atoi(gets(input));
+   gets(input);
+   Total_Particles = atoi(input);
    if (Total_Particles <= 0) {
       fprintf(stderr, "ERROR: The number of particles should be an int ");
       fprintf(stderr, "greater than 0.\n");
       fprintf(stderr, "If you need help, type \"nbody -help\".\n");
       exit(-1);
    }
-
-   Precision = atof(gets(input));
+   gets(input);
+   Precision = atof(input);
    if (Precision == 0.0) {
       fprintf(stderr, "ERROR: The precision has no default value.\n");
       fprintf(stderr, "If you need help, type \"nbody -help\".\n");
@@ -384,8 +384,8 @@
       fprintf(stderr, "If you need help, type \"nbody -help\".\n");
       exit(-1);
    }
-
-   Number_Of_Processors = atoi(gets(input));
+   gets(input);
+   Number_Of_Processors = atoi(input);
    if (Number_Of_Processors == 0) {
       fprintf(stderr, "ERROR: The Number_Of_Processors has no default.\n");
       fprintf(stderr, "If you need help, type \"nbody -help\".\n");
@@ -397,8 +397,8 @@
       fprintf(stderr, "If you need help, type \"nbody -help\".\n");
       exit(-1);
    }
-
-   Time_Steps = atoi(gets(input));
+   gets(input);
+   Time_Steps = atoi(input);
    if (Time_Steps == 0) {
       fprintf(stderr, "ERROR: The number of time steps has no default.\n");
       fprintf(stderr, "If you need help, type \"nbody -help\".\n");
@@ -410,8 +410,8 @@
       fprintf(stderr, "If you need help, type \"nbody -help\".\n");
       exit(-1);
    }
-
-   Timestep_Dur = atof(gets(input));
+   gets(input);
+   Timestep_Dur = atof(input);
    if (Timestep_Dur == 0.0) {
       fprintf(stderr, "ERROR: The duration of a time step has no default ");
       fprintf(stderr, "value.\n If you need help, type \"nbody -help\".\n");
@@ -423,8 +423,8 @@
       fprintf(stderr, "If you need help, type \"nbody -help\".\n");
       exit(-1);
    }
-
-   Softening_Param = atof(gets(input));
+   gets(input);
+   Softening_Param = atof(input);
    if (Softening_Param == 0.0)
       Softening_Param = MIN_REAL;
    if (Softening_Param < 0) {
