--- parsec-3.0-orig/pkgs/apps/blackscholes/src/blackscholes.c	2022-02-08 04:49:32.226930046 +0000
+++ parsec-3.0/pkgs/apps/blackscholes/src/blackscholes.c	2022-02-08 04:50:43.562390562 +0000
@@ -11,6 +11,7 @@
 #include <stdlib.h>
 #include <math.h>
 #include <string.h>
+#include "tern/user.h"
 
 #ifdef ENABLE_PARSEC_HOOKS
 #include <hooks.h>
@@ -437,6 +438,7 @@
     int *tids;
     tids = (int *) malloc (nThreads * sizeof(int));
 
+    slock_next_n(nThreads);
     for(i=0; i<nThreads; i++) {
         tids[i]=i;
         CREATE_WITH_ARG(bs_thread, &tids[i]);
