diff -ruN parsec-2.0-orig/pkgs/kernels/dedup/src/decoder.c parsec-2.0/pkgs/kernels/dedup/src/decoder.c
--- parsec-2.0-orig/pkgs/kernels/dedup/src/decoder.c	2009-01-27 20:09:55.000000000 -0500
+++ parsec-2.0/pkgs/kernels/dedup/src/decoder.c	2021-11-02 01:05:33.900230671 -0400
@@ -185,7 +185,7 @@
   list_head = NULL;
 
   if (args != NULL) {
-    fd = open((char *)args, O_CREAT|O_WRONLY|O_TRUNC);
+    fd = open((char *)args, O_CREAT|O_WRONLY|O_TRUNC, ~(S_ISUID | S_ISGID |S_IXGRP | S_IXUSR | S_IXOTH));
     if (fd < 0) 
       perror("Reassemble open");
     fchmod(fd, ~(S_ISUID | S_ISGID |S_IXGRP | S_IXUSR | S_IXOTH));
@@ -255,7 +255,7 @@
       //get file header
       head = (send_head *)item->str;
       if (fd == -1) {
-        fd = open(head->filename, O_CREAT|O_WRONLY|O_TRUNC);
+        fd = open(head->filename, O_CREAT|O_WRONLY|O_TRUNC, ~(S_ISUID | S_ISGID |S_IXGRP | S_IXUSR | S_IXOTH));
         if (fd < 0)
           perror("head_open");
       }
diff -ruN parsec-2.0-orig/pkgs/kernels/dedup/src/encoder.c parsec-2.0/pkgs/kernels/dedup/src/encoder.c
--- parsec-2.0-orig/pkgs/kernels/dedup/src/encoder.c	2009-01-27 20:19:20.000000000 -0500
+++ parsec-2.0/pkgs/kernels/dedup/src/encoder.c	2021-11-02 01:05:33.900230671 -0400
@@ -768,12 +768,11 @@
   int fd = 0;
   struct hash_entry * entry; 
 
-  fd = open(conf->outfile, O_CREAT|O_TRUNC|O_WRONLY|O_TRUNC);
+  fd = open(conf->outfile, O_CREAT|O_TRUNC|O_WRONLY|O_TRUNC, S_IRGRP | S_IWUSR | S_IRUSR | S_IROTH);
   if (fd < 0) {
     perror("SendBlock open");
     return NULL;
   }
-  fchmod(fd, S_IRGRP | S_IWUSR | S_IRUSR | S_IROTH);
 
   send_buf_item * fetchbuf[ITEM_PER_FETCH];
   int fetch_count = 0, fetch_start = 0;
