--- parsec-3.0-orig/pkgs/apps/raytrace/src/RTTL/common/RTThread.hxx	2022-04-18 22:04:14.023353330 +0000
+++ parsec-3.0/pkgs/apps/raytrace/src/RTTL/common/RTThread.hxx	2022-04-18 22:04:15.627347265 +0000
@@ -246,8 +246,8 @@
     pthread_cond_init(&m_cond, NULL);
   }
   ~MultiThreadedSyncPrimitive() {
-    pthread_mutex_destroy(&m_mutex);
-    pthread_cond_destroy(&m_cond);
+    // pthread_mutex_destroy(&m_mutex);
+    // pthread_cond_destroy(&m_cond);
   }
 
   // http://node1.yo-linux.com/cgi-bin/man2html?cgi_command=pthread_mutex_lock
